forc-pkg/src/pkg.rs

```diff
@@ -478,6 +493,65 @@ impl BuildPlan {
         Ok(plan)
     }
 
+    /// Return root packages in the order of required compilation/deployment order.
+    ///
+    /// Returns indices of nodes that are root packages given from the build plan graph.
+    /// To do so first gets `toposort` of the graph and filters the non-root members from the
+    /// resulting list of node indices.
+    pub fn root_pkgs(&self) -> Result<Vec<NodeIx>> {
+        let mut root_pkgs = Vec::new();
+        let graph = self.graph();
+        let parentless_nodes: HashSet<NodeIx> = parentless_nodes(graph).collect();
+        let mut pinned_to_root_map = HashMap::new();
+        // The graph will have roots of workspace members and pinned dependencies between them are
+        // explicitly inserted as a node in the graph. So a root node can have a corresponding
+        // pinned node which is a child of another root. While finding the order we need to take
+        // pinned versions into account so we should remove root version, if they have a
+        // corresponding pinned node in the graph. While doing so we should not lose root pkg
+        // indices that we removed for ordering since we are going to return corresponding root node
+        // indices as the ordered root_pkgs for this BuildPlan.
+        let mut ordering_graph = self.graph().clone();
+        for node in graph
+            .node_indices()
+            .filter(|node| matches!(graph[*node].source, SourcePinned::Root))
+        {
+            let root_version = &graph[node];
+            let pinned_version = graph
+                .node_indices()
+                .filter(|node| !matches!(graph[*node].source, SourcePinned::Root))
+                .find(|node| graph[*node].name == root_version.name);
+            if let Some(pinned_version) = pinned_version {
+                // Add this Root node as the corresponding node for `pinned_version`
+                pinned_to_root_map.insert(pinned_version, node);
+                // remove this root node from the ordering graph.
+                ordering_graph.remove_node(node);
+            }
+        }
```

Comment:
After looking a little closer, I see that it's not enough to check something like `is_parentless`, as workspace members that are depended on by another workspace will have a parent edge :thinking: 

Ideally, the graph should *never* have two packages for a single package. That is, member nodes that are depended on by some other node should never have a second equivalent package node in the graph - we should only have one.

I think there are a couple things we can do:

1. Update `fetch_deps` so that, when determining a dependency package's `Source`, before we assume it is a `Source::Path` we first check if the path is equal to a workspace `member` path. If so, then we add a `Source::Root` node with the desired `Path` instead.
2. On that note, we should probably change the `Source` and `SourcePinned` `Root` variant names to `Member`, as a `Root` implies the package has no incoming edges which is not always the case for member packages.

With these changes, we should be able to change this method to something like this:

```rust
    /// Produce an iterator yielding all workspace member nodes in order of compilation.
    ///
    /// In the case that this `BuildPlan` was constructed for a single package,
    /// only that package's node will be yielded.
    pub fn member_nodes(&self) -> '_ + impl Iterator<Item = NodeIx> {
        self.compilation_order()
            .iter()
            .cloned()
            .filter(|&n| self.graph[n].source == SourcePinned::Member)
    }
```